<a href="https://colab.research.google.com/github/susanaduque/Analitica-de-Negocios1/blob/main/arbol_de_decision_HD_Susana_Duque.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##*Caso de Estudio*

Una entidad médica desea implementar un modelo de árbol de decisión para pronosticar si las personas que hacen parte de la base de datos desarrollarán una enfermedad cardíaca.

Para este proceso utilizaremos las variables:

Edad: Indica la edad del paciente al momento del estudio.

Systolic: Presión arterial sistólica.

Diastolic: Presión arterial diastólica.

BMI: Índice de masa corporal del paciente.

La variable de salida será:

HeartDisease: Indica si la persona desarrolla enfermedad cardíaca (1 = Sí, 0 = No).

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1.Se procede con la carga de los datos

In [ ]:
nxl='/content/sample_data/2. HeartDisease (3).xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XD=XDB.iloc[:,[1,2,3,4]] #Variables de entrada
yd=XDB.iloc[:,0] #Variable de salida

2. Se procede con la implemetacion del de arbol

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd) #aqui el modelo busca la relacion entrada salida

#Y que fue lo que hizo el modelo?
ydp=mar.predict(XD) #esto es lo que pronostica el modelo

#Se construye la matrix de confusion
cm=confusion_matrix(yd,ydp)
display(cm)
VN=cm[0,0]; FP=cm[0,1];FN=cm[1,0]; VP=cm[1,1]

#Metricas de desempeño
Ex=(VP+VN)/(VP+VN+FP+FN) #Exactitud comportamiento general
print("la exactitud es: ",Ex)
Sen=VP/(VP+FN) #Sensibilidad: Como se comporta pronosticando PreAprobados
print("la sensibilidad es: ",Sen)
Esp=VN/(VN+FP) #Especificidad como se comporta frente a los negativos
print("la especificidad es: ",Esp)
Pre=VP/(VP+FP) #Precision como se comporta pronosticando los preaprobados
print("la precision es: ",Pre)
PreNeg=VN/(VN+FN) #Precision como se comporta pronosticando los negativos
print("la precision negativa es: ",PreNeg)

array([[63, 10],
       [20, 87]])

la exactitud es:  0.8333333333333334
la sensibilidad es:  0.8130841121495327
la especificidad es:  0.863013698630137
la precision es:  0.8969072164948454
la precision negativa es:  0.7590361445783133


3. despliegue el arbol de decisiones

In [ ]:
from sklearn.tree import export_graphviz #Exporta los datos a un grafico
from pydotplus import graph_from_dot_data #ese es un graficador

vs=["Edad","Systolic","Diastolic","BMI"] #titulos del arbol
dot_data=export_graphviz(mar,feature_names=vs)#Exportar de numeros a grafico en pdf
graph=graph_from_dot_data(dot_data) #Hacemos el grafico
graph.write_png('Arbol.png')


True

##*Análisis de Resultados*

El modelo de árbol de decisión presenta un desempeño adecuado en la clasificación de pacientes con y sin enfermedad cardíaca, alcanzando una exactitud del 83.33%, lo que indica un buen comportamiento general. La sensibilidad fue del 81.31%, evidenciando una correcta identificación de pacientes enfermos, mientras que la especificidad del 86.30% demuestra una adecuada clasificación de pacientes sanos. Asimismo, la precisión del 89.69% refleja que cuando el modelo predice la presencia de enfermedad cardíaca, acierta en casi 9 de cada 10 casos, mostrando una alta confiabilidad en sus predicciones positivas.

En cuanto a la estructura del árbol, la edad se identifica como la variable más influyente en la segmentación del riesgo. En pacientes mayores a 54.5 años se observa una proporción del 89.6% de casos positivos, generando un nodo con baja impureza (gini = 0.185), lo que indica una clasificación bastante homogénea. Además, en mayores de 69.5 años se presenta un nodo completamente puro (gini = 0), donde el 100% de los pacientes tiene enfermedad. En contraste, en el grupo de menor edad el comportamiento es más heterogéneo, por lo que variables como la presión arterial y el BMI permiten reducir la impureza y mejorar la clasificación. Por ejemplo, en pacientes jóvenes con presión sistólica mayor a 123 se identifica un nodo puro (gini = 0), evidenciando una regla clara de alto riesgo. En general, el modelo confirma que el riesgo cardiovascular aumenta principalmente con la edad, mientras que en pacientes más jóvenes depende en mayor medida de factores de presión arterial y metabólicos.

4. Analisis pacientes

In [ ]:
nxl='/content/sample_data/2. HeartDisease (3).xlsx'
XDB1_new = pd.read_excel(nxl, sheet_name=1)
XD_new_patients = XDB1_new.iloc[:, [1, 2, 3, 4]]
ydp_new_correct = mar.predict(XD_new_patients)
display(ydp_new_correct)
XD_new_patients['Prediction'] = ydp_new_correct
display(XD_new_patients)

array([0, 0, 0, 1, 1, 1, 1, 1])

,Age,Systolic,Diastolic,BMI,Prediction
0,33,120,64,33,0
1,55,115,81,32,0
2,51,130,70,20,0
3,65,101,63,18,1
4,62,109,110,38,1
5,60,130,69,35,1
6,61,124,95,36,1
7,53,150,81,41,1
